R/register_extendr.R

```diff
@@ -40,29 +45,51 @@ register_extendr <- function(path = ".", quiet = FALSE, force_wrappers = FALSE,
 
   outfile <- rprojroot::find_package_root_file("R", "extendr-wrappers.R", path = path)
 
-  # If force_wrappers is TRUE, use a minimal wrapper file even when
-  # make_wrappers() fails; since the wrapper generation depends on the compiled
-  # Rust code, the package needs to be installed before attempting this, but
-  # it's not always the case (e.g. the package might be corrupted, or not
-  # installed yet).
-  if (isTRUE(force_wrappers)) {
-    error_handle <- function(e) {
-      cli::cli_alert_danger("Failed to generate wrapper functions: {e$message}.")
-      cli::cli_alert_warning("Falling back to a minimal wrapper file instead.")
-
-      make_example_wrappers(pkg_name, outfile, path = path)
-    }
-  } else {
-    error_handle <- function(e) {
+  path <- rprojroot::find_package_root_file(path = path)
+
+  # If compile is NA, compile if the DLL is newer than the source files
+  if (isTRUE(is.na(compile))) {
+    compile <- needs_compilation(path, quiet) || pkgbuild::needs_compile(path)
+  }
+
+  if (isTRUE(compile)) {
+    # This relies on [`pkgbuild::needs_compile()`], which
+    # does not know about Rust files modifications.
+    # `force = TRUE` enforces compilation.
+    pkgbuild::compile_dll(
+      path = path,
+      force = TRUE,
+      quiet = quiet
+    )
+  }
+
+  library_path <- get_library_path(path)
+
+  if (!file.exists(library_path)) {
+    msg <- "{library_path} doesn't exist"
+    if (isTRUE(compile)) {
+      # If it doesn't exist even after compile, we have no idea what happened
+      ui_throw(msg)
+    } else {
+      # If compile wasn't invoked, it might succeed with explicit "compile = TRUE"
       ui_throw(
-        "Failed to generate wrapper functions.",
-        c(
-          ui_x(e[["message"]])
-        )
+        msg,
+        ui_i("You need to compile first, try {.code `register_rextendr(compile = TRUE)`}")
```

Comment:
Unfortunately, this will not work.
For some reason {cli} cannot be used directly to format strings, it always emits `message`s.
I believe this is by design but I would very much like to get an {cli}-interpolated `ansi_string`.
To avoid problems, remove `{.code ...}` and just write ``` `register_extendr(...)` ```.

If I find a solution, I will improve implementation of `ui_*` functions.

Reply:
Oh, thanks. I'm yet to figure out how to use these functions...